src/puppetlabs/puppetdb/query_eng.clj

```diff
@@ -182,14 +194,21 @@
     paging-options
     options :- query-options-schema
     row-fn]
-   (let [{:keys [scf-read-db url-prefix warn-experimental pretty-print]
+   (let [{:keys [scf-read-db url-prefix warn-experimental pretty-print log-queries]
           :or {warn-experimental true
-               pretty-print false}} options
+               pretty-print false
+               log-queries false}} options
+         log-id (when log-queries (.toString (java.util.UUID/randomUUID)))
          {:keys [remaining-query entity]} (eng/parse-query-context version query warn-experimental)
          munge-fn (get-munge-fn entity version paging-options url-prefix)]
+
+     (when log-queries
+       ;; log the AST of the incoming query
```

Comment:
Do we want a way to differentiate between queries submitted externally vs. those submitted via the TK PuppetDBServer service's query method when logging?

Reply:
I don't think it matters, we want to log the queries regardless of where they come from. But I think the only places the TK service method is used, for better or for worse, is in HA sync and the admin export comand.